### ISSUE TITLE
Add support for processing instructions (`<?...?>`)

### ISSUE THREAD

**Issue by aJanuary** [chrisbottin/xml-formatter#12]
I would like it to handle processing instructions (things with `<? ... ?>` such as `<?xml-stylesheet ...?>`) Currently the declarations cause it to finish parsing early and the output is truncated.

Input:
```<!-- Load the stylesheet --><?xml-stylesheet href="style.xsl" type="text/xsl" ?><foo></foo>```

Desired parse output:
```
{
  children: [
    { content: '<!-- Load the stylesheet -->', name: '#comment' },
    { content: ' href="style.xsl" type="text/xsl" ', target: 'xml-stylesheet', name: '#processingInstruction' },
    { attributes: {}, children: [], name: 'foo' }
  ],
  declaration: undefined,
  root: { attributes: {}, children: [], name: 'foo' }
}
```

Desired formatted output:
```
<!-- Load the stylesheet -->
<?xml-stylesheet href="style.xsl" type="text/xsl"?>
<foo>
</foo>
```

Current parse output:
```
{
  children: [
    { content: '<!-- Load the stylesheet -->', name: '#comment' }
  ],
  declaration: undefined,
  root: undefined
}
```

Current formatted output:
```
<!-- Load the stylesheet -->
```

**Comment by aJanuary** [comment]
I have changed the library for my own purposes so that it handles processing instructions, and will happily turn that into a PR (one for xml-formatter and one for xml-parser-xo). However, it also changes how non-significant whitespace is handled, and I don't know if it gels with the trim argument.

It seems currently the way of handling whitespace outside of the root is to trim the string if the trim option is true. This allows it to handle strings like `\n\n<foo></foo>\n\n\n`. However, processing instructions can also go in the root of the document e.g. `<?xml-stylesheet href="style.xsl" type="text/xsl" ?>\n<foo></foo>\n\n\n`? This means there can be whitespace between the processing instruction and the next node that needs to be consumed and ignored. My PR would allow multiple comments and processing instructions at the root, while consuming and throwing away any whitespace between them.

```
<!-- Load the stylesheet -->
<?xml-stylesheet href="style.xsl" type="text/xsl" ?>

<foo>
</foo>

```

would be parsed as

```
{
  children: [
    { content: '<!-- Load the stylesheet -->', name: '#comment' },
    { content: ' href="style.xsl" type="text/xsl" ', target: 'xml-stylesheet', name: '#processingInstruction' },
    { attributes: {}, children: [], name: 'foo' }
  ],
  declaration: undefined,
  root: { attributes: {}, children: [], name: 'foo' }
}
```

which seems fine.

However, because of the new behaviour of ignoring non-significant whitespace, `var node = parse('   <foo></foo>   ', {trim: false});` would parse fully instead of returning undefined, which is a change to the existing "it should not trim the input when option is false" test. It kind of renders the whole trim option meaningless.

One solution would be to slightly change the meaning of the trim argument to go from "strip the whitespace before and after" to "ignore non-significant whitespace". It makes the name of the argument slightly odd, but it maintains a higher level of compatibility with the existing behaviour (it would still change the behaviour such that `<!-- Comment 1 -->\n<!-- Comment 2 -->\n<Foo></Foo>` parses fully rather than stopping at the end of the first comment)

Do you have any input a) on whether such a PR is likely to be accepted and b) how to handle the non-significant whitespace/trim issue?

**Comment by chrisbottin** [comment]
Thanks for your messages @aJanuary , I will review your comments and get back to you shortly.

**Comment by chrisbottin** [comment]
Hey @aJanuary,

I've read your comments, first of all thank for willing to contribute to these repositories.

I think your proposed changes make sense and we should remove the current options in `xml-parser-xo` so the non-significant whitespaces are always ignored and the comments are  always preserved.

The eventual goal is to be able to load any valid XML content such as something like:

```
<?xml version="1.0" encoding="utf-8"?>
<!-- Load the stylesheet -->
<?xml-stylesheet href="style.xsl" type="text/xsl" ?>
<!DOCTYPE ExampleType SYSTEM "example.dtd">
<foo><![CDATA[some text]]> and some more</foo>
```

As it stands `<!DOCTYPE` and `<![CDATA` are also not supported as per the other two opened issues.

If you also want to take a look at adding the support for those that would be great, otherwise I will after your PR.

Many Thanks,

Chris

**Comment by chrisbottin** [comment]
This should be fixed by #15
